ext/standard/hrtime.c

```diff
@@ -0,0 +1,206 @@
+	Public Domain - 2011 Mattias Jansson / Rampant Pixels */
+
+#if PHP_HRTIME_PLATFORM_POSIX
+
+# include <unistd.h>
+# include <time.h>
+# include <string.h>
+
+#elif PHP_HRTIME_PLATFORM_WINDOWS
+
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+
+static uint64_t _timer_freq = 0;
+static double _timer_int = .0;
+
+#elif PHP_HRTIME_PLATFORM_APPLE
+
+# include <mach/mach_time.h>
+# include <string.h>
+static mach_timebase_info_data_t _timerlib_info;
+
+#elif PHP_HRTIME_PLATFORM_HPUX
+
+# include <sys/time.h>
+
+#elif PHP_HRTIME_PLATFORM_AIX
+
+# include <sys/time.h>
+# include <sys/systemcfg.h>
+
+#endif
+
+#define NANO_IN_SEC 1000000000
+/* }}} */
+
+static int _timer_init()
+{/*{{{*/
+#if PHP_HRTIME_PLATFORM_WINDOWS
+
+	uint64_t unused;
+	if (!QueryPerformanceFrequency((LARGE_INTEGER*) &_timer_freq) ||
+		!QueryPerformanceCounter((LARGE_INTEGER*) &unused)) {
+		return -1;
+	}
+	_timer_int = 1.0 / _timer_freq;
+
+#elif PHP_HRTIME_PLATFORM_APPLE
+
+	if (mach_timebase_info(&_timerlib_info)) {
+		return -1;
+	}
+
+#elif PHP_HRTIME_PLATFORM_POSIX
+
+	struct timespec ts = { .tv_sec = 0, .tv_nsec = 0 };
+
+	if (clock_gettime(CLOCK_MONOTONIC, &ts)) {
+		return -1;
+	}
+
+#elif PHP_HRTIME_PLATFORM_HPUX
+
+	/* pass */
+
+#elif PHP_HRTIME_PLATFORM_AIX
+
+	/* pass */
+
+#else
+	/* Timer unavailable. */
+	return -1;
+#endif
+
+	return 0;
+}/*}}}*/
+
+/* {{{ */
+PHP_MINIT_FUNCTION(hrtime)
+{
+	if (0 > _timer_init()) {
+		php_error_docref(NULL, E_WARNING, "Failed to initialize high-resolution timer");
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+/* }}} */
+
+static zend_always_inline php_hrtime_t _timer_current(void)
+{/*{{{*/
+#if PHP_HRTIME_PLATFORM_WINDOWS
+	uint64_t cur;
+	QueryPerformanceCounter((LARGE_INTEGER*) &cur);
+	return (php_hrtime_t)((double)cur * _timer_int * (double)NANO_IN_SEC);
+#elif PHP_HRTIME_PLATFORM_APPLE
+	return (php_hrtime_t)mach_absolute_time() * _timerlib_info.numer / _timerlib_info.denom;
+#elif PHP_HRTIME_PLATFORM_POSIX
+	struct timespec ts = { .tv_sec = 0, .tv_nsec = 0 };
+	if (0 == clock_gettime(CLOCK_MONOTONIC, &ts)) {
+		return ((php_hrtime_t) ts.tv_sec * (php_hrtime_t)NANO_IN_SEC) + ts.tv_nsec;
+	}
+	return 0;
+#elif PHP_HRTIME_PLATFORM_HPUX
+	return (php_hrtime_t) gethrtime();
+#elif  PHP_HRTIME_PLATFORM_AIX
+	timebasestruct_t t;
+	read_wall_time(&t, TIMEBASE_SZ);
+	time_base_to_time(&t, TIMEBASE_SZ);
+	return (php_hrtime_t) t.tb_high * (php_hrtime_t)NANO_IN_SEC + t.tb_low;
+#else
+	return 0;
+#endif
+}/*}}}*/
+
+#if ZEND_ENABLE_ZVAL_LONG64
+#define PHP_RETURN_HRTIME(t) RETURN_LONG((zend_long)t)
+#else
+#ifdef _WIN32
+# define HRTIME_U64A(i, s, len) _ui64toa_s(i, s, len, 10)
+#else
+# define HRTIME_U64A(i, s, len) \
+	do { \
+		int st = snprintf(s, len, "%llu", i); \
+		s[st] = '\0'; \
+	} while (0)
+#endif
+#define PHP_RETURN_HRTIME(t) do { \
+	char _a[ZEND_LTOA_BUF_LEN]; \
+	double _d; \
+	HRTIME_U64A(t, _a, ZEND_LTOA_BUF_LEN); \
+	_d = zend_strtod(_a, NULL); \
```

Comment:
I'm not sure why we're converting to string and back here - can't we just do `(double)`? Yes, there's a potential for precision loss, but wouldn't it happen in string case too?

Reply:
The integral part of double can be overflown quite fast. With an up time of a couple of months, as soon as timestamp exceeds `(2^52)/1000000000`. The string conversion is the only way i see so far to do the correct conversion. Still it's 32-bit only. On 64-bit, even a signed 64-bit integral is fine for some hundreds of years of up time.